### PR TITLE
[py2py3] WMCore/Algorithms - migration

### LINKED ISSUE #10009: [py2py3]  Migrate WMCore/Algorithms
**Impact of the new feature**

This involves migrating WMCore.Algorithms All its dependencies have already been migrated.

**Is your feature request related to a problem? Please describe.**

Make WMCore.Algorithms compatible with both py2 and py3.

**Describe the solution you'd like**

Run futurize on WMCore.Algorithms and then apply manual changes, if necessary.

### PR DESCRIPTION
**Pull request by mapellidario**
Fixes #10009 

#### Status

- src: ready
- test: not necessary

#### Description

Run futurize on WMCore/Algorithms

Changes:

str: `from builtins import str` used in
- src/python/WMCore/Algorithms/MathAlgos.py: used in `decimal.Decimal(str(stdDev)).is_finite()`
- src/python/WMCore/Algorithms/SubprocessAlgos.py

list and iterators: range
- src/python/WMCore/Algorithms/MathAlgos.py

[Iterate over dictionariy keys](http://python-future.org/compatible_idioms.html#iterating-through-dict-keys-values-items) used in:
- src/python/WMCore/Algorithms/MiscAlgos.py

[iterate over dictionary items](http://python-future.org/compatible_idioms.html#iterating-through-dict-keys-values-items) used in:
- src/python/WMCore/Algorithms/ParseXMLFile.py

oct: `from builtins import oct`
- src/python/WMCore/Algorithms/Permissions.py

#### Is it backward compatible (if not, which system it affects?)

It should be. However, any bug would affect directly, 

```
'WMCore_Algorithms': ['WMComponent_TaskArchiver',
                       'WMCore_DataStructs',
                       'WMComponent_DBS3Buffer',
                       'WMCore_FwkJobReport',
                       'WMCore_WorkQueue',
                       'WMCore_ACDC',
                       'WMCore_Algorithms',
                       'WMCore_WMSpec',
                       'WMCore_Services',
                       'WMCore_Storage',
                       'WMCore_WMRuntime'],
```

#### External dependencies / deployment changes

requires python-future

### CODE REVIEW

**Comment by mapellidario** [comment]
The suggested fix for newer python3 idioms, instead of pre-py2.6 ones, is the usual check with `type()` and not `isinstance`.

```python
--- src/python/WMCore/Algorithms/SubprocessAlgos.py	(original)
+++ src/python/WMCore/Algorithms/SubprocessAlgos.py	(refactored)
@@ -100,7 +100,7 @@
     if timeout:
-        if type(timeout) != int:
+        if not isinstance(timeout, int):
```

I can try and change it, it seems safe enough

File: src/python/WMCore/Algorithms/MiscAlgos.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def sortListByKey(data, key):
                 logging.error("Found list entry with empty key set in sortListByKey: %s" % entry)
                 logging.error("Skipping")
                 continue
-        if not value in final.keys():
+        if not value in final:
```
Comment (amaltaro) [inline src/python/WMCore/Algorithms/MiscAlgos.py:38]:
```suggestion
        if value not in final:
```

File: src/python/WMCore/Algorithms/ParseXMLFile.py [hunk 2]
```diff
@@ -28,7 +31,7 @@ def __init__(self, name, attrs):
         self.name = str(name)
         self.attrs = {}
         self.text = None
-        [ self.attrs.__setitem__(str(k), str(v)) for k,v in attrs.items()]
+        [ self.attrs.__setitem__(str(k), str(v)) for k,v in viewitems(attrs)]
```
Comment (amaltaro) [inline src/python/WMCore/Algorithms/ParseXMLFile.py:34]:
```suggestion
        for k, v in viewitems(attrs):
            self.attrs.__setitem__(str(k), str(v))
```
Please check the spaces/indentantion though.

**Review by amaltaro — CHANGES_REQUESTED** [review]

**Comment by amaltaro** [comment]
(Pushed the submit button before writing my final comments, sorry).
@mapellidario I have made some minor comments that would be useful to be taken into account.

I know it's out of the scope of these python3 changes, but could you please also fix those warnings (bold) reported in the jenkins summary (only the simple and safe ones), like:
```
src/python/WMCore/Algorithms/MathAlgos.py
 W1201, line 56 in getAverageStdDev: Specify string format arguments as logging function parameters
 W0622, line 141 in createHistogram: Redefining built-in 'bin' 
...
```

File: src/python/WMCore/Algorithms/ParseXMLFile.py (hunk 2 above)
Comment (todor-ivanov) [reply src/python/WMCore/Algorithms/ParseXMLFile.py:34]: Agree with the proposed change here!

**Review by todor-ivanov — APPROVED** [review]
Looks good to me.

File: src/python/WMCore/Algorithms/MiscAlgos.py (hunk 1 above)
Comment (mapellidario) [reply src/python/WMCore/Algorithms/MiscAlgos.py:38]: Why do I keep not noticing this? it's like the 4th time you suggest this change to me! Thanks!

File: src/python/WMCore/Algorithms/ParseXMLFile.py (hunk 2 above)
Comment (mapellidario) [reply src/python/WMCore/Algorithms/ParseXMLFile.py:34]: This is definitely better, Thanks!

**Comment by mapellidario** [comment]
> I know it's out of the scope of these python3 changes, but could you please also fix those warnings (bold) reported in the jenkins summary (only the simple and safe ones), like:
> 

I addressed all the warnings from pylint in bold but one: it complains that the process defined at [this line](https://github.com/dmwm/WMCore/blob/870d9179bdf07373c4fd7e4027ce0e53605c68dd/src/python/WMCore/Algorithms/SubprocessAlgos.py#L69) is not assigned to anything. It does not seem that we are using any information coming from that process, nor we check its status, so I did not change that line. Would you prefer a different solution?

**Comment by amaltaro** [comment]
> ...It does not seem that we are using any information coming from that process, nor we check its status, so I did not change that line. Would you prefer a different solution?

No, we can leave it alone. Thanks Dario.

**Comment by mapellidario** [comment]
Perfect, thanks! Then I consider this ready to be merged (with absolutely no hurry at all!).

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Review by amaltaro — APPROVED** [review]

**Comment by amaltaro** [comment]
Thanks Dario!

**Pull request merged**